### ISSUE TITLE
Parsing at the tip of the document in CoqIDE

### ISSUE THREAD

**Issue by Tuplanolla** [coq/coq#9385]
#### Version

Coq 8.8.2

#### Operating system

Red Hat Enterprise Linux Workstation 7.4

#### Description of the problem

The following actions prevent further navigation forward.

* Open CoqIDE.
* Insert `Debug On.` into the buffer.
* Navigate forward.
* Ignore the error message.
* Insert `Set` before `Debug`.
* Replace `On` with `Unification`.
* Navigate forward.

```
Currently, the parsing api only supports parsing at the tip of the document.
You wanted to parse at: 2 but the current tip is: 3
```

The only way to recover is to navigate backward.

#### Notes

This is not a duplicate of #7142,
because triggering this bug does not require first triggering an anomaly.

**Comment by ejgallego** [comment]
Can you reproduce with `master`? This will likely fixed in 8.10 [if not fixed already]

**Comment by gares** [comment]
The bug is here:
  https://github.com/coq/coq/blob/master/ide/idetop.ml#L90
the code calls `Stm.add` then decides to signal an error but does not tell the `Stm` that the error is there (it does not backtrack the sentence). So the document gets out of sync, Coq thinks the sentence is there but the UI does not.
